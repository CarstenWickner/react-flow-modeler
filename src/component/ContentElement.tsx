import * as React from "react";

import { HorizontalStroke } from "./HorizontalStroke";

export class ContentElement extends React.Component<{
    elementId: string;
    editMenu: (() => React.ReactNode) | undefined;
    onSelect: (elementId: string) => void;
}> {
    onClick = (event: React.MouseEvent): void => {
        const { onSelect, elementId } = this.props;
        onSelect(elementId);
        event.stopPropagation();
    };

    render(): React.ReactNode {
        const { editMenu, children } = this.props;
        return (
            <>
                <div className="stroke-horizontal arrow" />
                <div className="flow-element-wrapper">
                    <div className={`flow-element content-element${editMenu ? " selected" : ""}`} onClick={this.onClick}>
                        {children}
                    </div>
                    {editMenu && editMenu()}
                </div>
                <HorizontalStroke optional />
            </>
        );
    }
}
