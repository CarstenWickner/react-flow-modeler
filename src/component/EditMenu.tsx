import * as React from "react";

import { ElementType } from "../types/GridCellData";
import { FlowModelerProps, MenuOptions } from "../types/FlowModelerProps";

export class EditMenu extends React.Component<{
    targetType:
        | ElementType.Start
        | ElementType.Content
        | ElementType.GatewayDiverging
        | ElementType.GatewayConverging
        | ElementType.ConnectGatewayToElement;
    referenceElementId: string;
    menuOptions?: FlowModelerProps["options"]["editActions"];
}> {
    onClick = (event: React.MouseEvent): void => event.stopPropagation();

    renderMenuItem = (options: MenuOptions, defaultClassName: string): React.ReactNode => {
        const { targetType, referenceElementId } = this.props;
        return (
            (!options || !options.isActionAllowed || options.isActionAllowed(targetType, referenceElementId)) && (
                <span title={options && options.title} className={(options && options.className) || defaultClassName} />
            )
        );
    };

    render(): React.ReactNode {
        const { targetType, menuOptions } = this.props;
        const { addDivergingBranch, addFollowingContentElement, addFollowingDivergingGateway, changeNextElement, removeElement } = menuOptions || {};
        return (
            <div className="menu" onClick={this.onClick}>
                {targetType === ElementType.GatewayDiverging && this.renderMenuItem(addDivergingBranch, "menu-item add-branch")}
                {targetType !== ElementType.GatewayDiverging && this.renderMenuItem(addFollowingContentElement, "menu-item add-content")}
                {targetType !== ElementType.GatewayDiverging && this.renderMenuItem(addFollowingDivergingGateway, "menu-item add-gateway")}
                {targetType !== ElementType.GatewayDiverging &&
                    targetType !== ElementType.GatewayConverging &&
                    this.renderMenuItem(changeNextElement, "menu-item change-next")}
                {targetType !== ElementType.Start &&
                    targetType !== ElementType.GatewayConverging &&
                    this.renderMenuItem(removeElement, "menu-item remove")}
            </div>
        );
    }
}
