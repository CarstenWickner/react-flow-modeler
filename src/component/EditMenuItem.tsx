import * as React from "react";
import { useDrag } from "react-dnd";

import { DraggedLinkContext } from "../types/EditAction";

const disabledDragging: [{ isDragging: boolean }, React.LegacyRef<HTMLSpanElement>] = [{ isDragging: false }, undefined];

export const EditMenuItem: React.FC<{
    options: { className?: string; title?: string };
    defaultClassName: string;
    onClick?: (event: React.MouseEvent) => void;
    dragItem?: DraggedLinkContext;
}> = ({ options, defaultClassName, onClick, dragItem }) => {
    if (!dragItem && !onClick) {
        return null;
    }
    const [{ isDragging }, drag] = dragItem
        ? useDrag({
              item: dragItem,
              collect: (monitor) => ({ isDragging: !!monitor.isDragging() })
          })
        : disabledDragging;
    const baseClassName = (options && options.className) || `menu-item ${defaultClassName}`;
    return (
        <span
            className={`${baseClassName}${isDragging ? " dragging" : ""}`}
            {...{ title: options ? options.title : undefined, ref: drag, onClick }}
        />
    );
};
