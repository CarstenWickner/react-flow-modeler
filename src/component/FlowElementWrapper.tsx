import * as React from "react";
import { useDrop } from "react-dnd";

import { isFlowValid } from "../model/pathValidationUtils";

import { ContentNode, ConvergingGatewayNode, DivergingGatewayNode, ElementType, EndNode, ModelElementExclStart } from "../types/ModelElement";
import { DraggableType, DraggedLinkContext, onLinkDropCallback } from "../types/EditAction";

const isTargetAncestorOfDragItem = (
    originElement: ModelElementExclStart,
    referenceElement: ContentNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode
): boolean => {
    if (originElement === referenceElement) {
        return true;
    }
    if (originElement.type === ElementType.ConvergingGatewayNode) {
        return originElement.precedingBranches.some((branch) => isTargetAncestorOfDragItem(branch, referenceElement));
    }
    return (
        originElement.precedingElement.type !== ElementType.StartNode && isTargetAncestorOfDragItem(originElement.precedingElement, referenceElement)
    );
};

const isDropValid = (referenceElement: ContentNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode, onDrop: onLinkDropCallback) => (
    dragContext: DraggedLinkContext
): boolean =>
    !isTargetAncestorOfDragItem(dragContext.originElement, referenceElement) &&
    dragContext.originElement.followingElement !== referenceElement &&
    isFlowValid(onDrop(referenceElement, dragContext, true).changedFlow);

const disabledDropping: [{ isOver: boolean; canDrop: false }, React.LegacyRef<HTMLDivElement>] = [{ isOver: false, canDrop: false }, undefined];

export const FlowElementWrapper: React.FC<{
    elementTypeClassName: string;
    referenceElement: ContentNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode;
    editMenu?: (() => React.ReactNode) | undefined;
    onLinkDrop?: onLinkDropCallback | undefined;
    onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;
}> = ({ elementTypeClassName, referenceElement, editMenu, onLinkDrop, onClick, children }) => {
    const [{ isOver, canDrop }, drop] = onLinkDrop
        ? useDrop({
              accept: DraggableType.LINK,
              canDrop: isDropValid(referenceElement, onLinkDrop),
              drop: (dragContext: DraggedLinkContext) => onLinkDrop(referenceElement, dragContext),
              collect: (monitor) => ({
                  isOver: !!monitor.isOver(),
                  canDrop: !!monitor.canDrop()
              })
          })
        : disabledDropping;
    return (
        <>
            <div className={`stroke-horizontal arrow${isOver && canDrop ? " can-drop" : ""}`} />
            <div className={`flow-element-wrapper${isOver ? (canDrop ? " can-drop" : " cannot-drop") : ""}`} ref={drop}>
                <div className={`flow-element ${elementTypeClassName}${editMenu ? " selected" : ""}`} onClick={onClick}>
                    {children}
                </div>
                {editMenu && editMenu()}
            </div>
        </>
    );
};
