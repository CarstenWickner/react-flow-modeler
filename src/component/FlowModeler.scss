@import "node_modules/mathsass/dist/math";

.flow-modeler {
    display: grid;
    padding: 1em;
    align-content: stretch;
    justify-content: stretch;

    $strokeColor: #01579b;
    $strokeStrength: 0.25em;
    $arrowSize: 1.5 * $strokeStrength;
    $strokeLength: 1.75em;
    $elementSizeFactor: 2.5;
    $standardElementSize: $elementSizeFactor * 1em;
    $gatewayBorderSize: sqrt($elementSizeFactor * $elementSizeFactor * 2) * 1em / 2;
    $standardElementColor: #808080;
    $selectionColor: #ffab00;

    & .grid-cell {
        display: flex;
        align-items: center;
        align-content: center;
        min-height: $strokeLength;
        position: relative;

        & .flow-element {
            display: flex;
            position: relative;
            background-color: white;
            margin: $strokeStrength 0;

            &.start-element {
                border: ($strokeStrength / 2) solid $standardElementColor;
                height: $standardElementSize - $strokeStrength;
                width: $standardElementSize - $strokeStrength;
                border-radius: 50%;
            }
            &.end-element {
                box-shadow: 0 0 0 ($strokeStrength * 0.9) $standardElementColor;
                border: ($strokeStrength * 0.9) solid white;
                background-color: $standardElementColor;
                height: $standardElementSize - (4 * 0.9 * $strokeStrength);
                width: $standardElementSize - (4 * 0.9 * $strokeStrength);
                border-radius: 50%;
                margin: $strokeStrength ($strokeStrength * 0.9);

                &.selected {
                    box-shadow: 0 0 0 ($strokeStrength * 0.9) $selectionColor;
                }
            }
            &.content-element {
                border: ($strokeStrength / 3) solid $standardElementColor;
                border-radius: 0.25em;
                min-width: $standardElementSize;
                min-height: $standardElementSize / 2;
            }
            &.gateway-element {
                position: relative;
                height: $standardElementSize;
                width: $standardElementSize;
                margin: 0 $strokeStrength / 2;
                z-index: 1;

                display: flex;
                align-items: center;
                justify-content: center;

                &::before {
                    content: "";
                    position: absolute;
                    transform: rotate(45deg);
                    border: ($strokeStrength / 2) solid $standardElementColor;
                    height: $gatewayBorderSize;
                    width: $gatewayBorderSize;
                    background-color: transparent;
                }
                &::after {
                    content: "X";
                    position: absolute;
                    font: bold 1.75em Arial, sans-serif;
                    color: $standardElementColor;
                }
            }
            &.start-element.selected,
            &.content-element.selected,
            &.gateway-element.selected::before {
                border-color: $selectionColor;
            }
        }
        & .menu {
            position: absolute;
            display: flex;
            flex-direction: column;
            align-items: center;
            left: calc(100% + #{$strokeStrength});
            width: 1em;
            margin: auto;
            border: ($strokeStrength / 3) solid $selectionColor;
            border-radius: 0.25em;
            background-color: white;
            z-index: 1;
            padding: $strokeStrength;
        }
        & .stroke-horizontal::before,
        & .stroke-vertical::before {
            background-color: $strokeColor;
        }
        & .stroke-horizontal.selected::before {
            background-color: $selectionColor;
        }
        & .stroke-horizontal {
            position: relative;
            display: flex;
            flex-direction: column;
            flex-grow: 1;
            min-width: $strokeLength;

            &.optional {
                min-width: 0;
            }
            &.arrow {
                margin-right: $arrowSize;
        
                &::after {
                    content: "";
                    height: 0;
                    width: 0;
                    position: absolute;
                    right: 0;
                    top: calc(50% - #{$arrowSize});
                    margin-right: -$arrowSize;
                    border-top: $arrowSize solid transparent;
                    border-left: $arrowSize solid $strokeColor;
                    border-bottom: $arrowSize solid transparent;
                }
            }
            &::before {
                position: absolute;
                content: "";
                height: $strokeStrength;
                width: 100%;
                top: calc(50% - #{$strokeStrength / 2});
            }
            & > .top-label {
                display: flex;
                flex-direction: column;
                justify-content: flex-end;
                align-content: center;
                padding: 0 0.25em;
                margin-bottom: $strokeStrength / 2;
                min-height: $strokeStrength * 4;
            }
            & > .bottom-spacing {
                margin-top: $strokeStrength / 2;
            }
        }
        & .stroke-vertical {
            position: relative;
            display: flex;
            flex-direction: column;
            align-self: stretch;
            width: $strokeStrength;
            &::before {
                position: absolute;
                content: "";
                width: $strokeStrength;
            }
            &.top-half::before {
                height: calc(50% + #{$strokeStrength / 2});
                top: 0;
            }
            &.full-height::before {
                height: 100%;
                top: 0;
            }
            &.bottom-half::before {
                height: calc(50% + #{$strokeStrength / 2});
                bottom: 0;
            }
        }
    }
    & .menu-item {
        font-family: Arial, Helvetica, sans-serif;
        line-height: 1.1em;

        &.add-branch::before {
            content: "↳";
        }
        &.add-content::before {
            content: "■";
        }
        &.add-gateway::before {
            content: "◆";
        }
        &.change-next::before {
            content: "⇥";
        }
        &.remove::before {
            content: "×";
        }
    }
}
