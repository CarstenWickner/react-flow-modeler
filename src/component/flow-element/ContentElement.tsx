import * as React from "react";

export const ContentElement: React.FunctionComponent<{
    children: React.ReactChild;
}> = ({ children }) => (
    <>
        <div className="arrow" />
        <div className="flow-element content-element">{children}</div>
    </>
);
