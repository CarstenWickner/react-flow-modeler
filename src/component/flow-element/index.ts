export { Start } from "./Start";
export { ContentElement } from "./ContentElement";
export { Gateway } from "./Gateway";
export { GatewayConnector } from "./GatewayConnector";
export { StrokeExtension } from "./StrokeExtension";
export { End } from "./End";
