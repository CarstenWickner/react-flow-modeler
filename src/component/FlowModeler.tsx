import * as PropTypes from "prop-types";
import * as React from "react";

import { ContentElement } from "./ContentElement";
import { Gateway } from "./Gateway";
import { HorizontalStroke } from "./HorizontalStroke";
import { EditMenu } from "./EditMenu";
import { GridCell } from "./GridCell";
import { buildRenderData } from "./renderDataUtils";
import { GridCellData, ElementType } from "../types/GridCellData";
import { FlowModelerProps } from "../types/FlowModelerProps";

import "./FlowModeler.scss";

type SelectableElementType =
    | ElementType.Start
    | ElementType.Content
    | ElementType.GatewayDiverging
    | ElementType.GatewayConverging
    | ElementType.ConnectGatewayToElement;

interface FlowModelerState {
    selectedType: SelectableElementType;
    selectedId?: string;
}

export class FlowModeler extends React.Component<FlowModelerProps, FlowModelerState> {
    state: FlowModelerState = { selectedType: null, selectedId: null };

    isTargetSelected(type: SelectableElementType, id?: string): boolean {
        return this.state.selectedType === type && this.state.selectedId === id;
    }

    onSelect = (selectedType: SelectableElementType, selectedId?: string): void => {
        const { options } = this.props;
        if ((!options || !options.readOnly) && !this.isTargetSelected(selectedType, selectedId)) {
            this.setState({
                selectedType,
                selectedId
            });
        }
    };

    onClickStart = (event: React.MouseEvent): void => {
        this.onSelect(ElementType.Start);
        event.stopPropagation();
    };

    onSelectContent = (elementId: string): void => this.onSelect(ElementType.Content, elementId);

    onSelectGatewayDiverging = (gatewayId: string): void => this.onSelect(ElementType.GatewayDiverging, gatewayId);

    onSelectGatewayConverging = (followingElementId?: string): void => this.onSelect(ElementType.GatewayConverging, followingElementId);

    onSelectConnectGatewayToElement = (followingElementId?: string): void => this.onSelect(ElementType.ConnectGatewayToElement, followingElementId);

    clearSelection = (event: React.MouseEvent): void => {
        this.onSelect(ElementType.Content, null);
        event.stopPropagation();
    };

    renderEditMenu = (targetType: SelectableElementType, referenceElementId?: string): React.ReactNode | undefined => {
        if (!this.isTargetSelected(targetType, referenceElementId)) {
            return undefined;
        }
        const { options } = this.props;
        const menuOptions = options ? options.editActions : undefined;
        return <EditMenu targetType={targetType} referenceElementId={referenceElementId} menuOptions={menuOptions} />;
    };

    renderFlowElement(cellData: GridCellData): React.ReactNode {
        switch (cellData.type) {
            case ElementType.Start:
                const startEditMenu = this.renderEditMenu(cellData.type);
                return (
                    <>
                        <div className={`flow-element start-element${startEditMenu ? " selected" : ""}`} onClick={this.onClickStart} />
                        {startEditMenu}
                    </>
                );
            case ElementType.Content:
                const { renderContent } = this.props;
                const contentEditMenu = this.renderEditMenu(cellData.type, cellData.elementId);
                return (
                    <ContentElement elementId={cellData.elementId} editMenu={contentEditMenu} onSelect={this.onSelectContent}>
                        {renderContent({
                            elementData: cellData.data,
                            contentElementId: cellData.elementId
                        })}
                    </ContentElement>
                );
            case ElementType.GatewayDiverging:
                const { renderGatewayConditionType } = this.props;
                const divergingGatewayEditMenu = this.renderEditMenu(cellData.type, cellData.gatewayId);
                return (
                    <Gateway
                        type={cellData.type}
                        gatewayId={cellData.gatewayId}
                        editMenu={divergingGatewayEditMenu}
                        onSelect={this.onSelectGatewayDiverging}
                    >
                        {renderGatewayConditionType &&
                            renderGatewayConditionType({
                                gatewayData: cellData.data,
                                gatewayElementId: cellData.gatewayId
                            })}
                    </Gateway>
                );
            case ElementType.GatewayConverging:
                const convergingGatewayEditMenu = this.renderEditMenu(cellData.type, cellData.followingElementId);
                return (
                    <Gateway
                        type={cellData.type}
                        followingElementId={cellData.followingElementId}
                        editMenu={convergingGatewayEditMenu}
                        onSelect={this.onSelectGatewayConverging}
                    />
                );
            case ElementType.ConnectGatewayToElement:
                const { renderGatewayConditionValue } = this.props;
                const branchElementId = cellData.elementId;
                const connectorEditMenu = this.renderEditMenu(cellData.type, branchElementId);
                return (
                    <HorizontalStroke
                        incomingConnection={cellData.connectionType}
                        followingElementId={branchElementId}
                        editMenu={connectorEditMenu}
                        onSelect={this.onSelectConnectGatewayToElement}
                    >
                        {renderGatewayConditionValue &&
                            renderGatewayConditionValue({
                                conditionData: cellData.data,
                                gatewayElementId: cellData.gatewayId,
                                branchElementId
                            })}
                    </HorizontalStroke>
                );
            case ElementType.ConnectElementToGateway:
                return <HorizontalStroke outgoingConnection={cellData.connectionType} />;
            case ElementType.StrokeExtension:
                return <div className="stroke-horizontal optional" />;
            case ElementType.End:
                return (
                    <>
                        <div className="stroke-horizontal arrow" />
                        <div className="flow-element end-element" />
                    </>
                );
        }
    }

    renderGridCell = (): ((cellData: GridCellData) => React.ReactNode) => {
        const { options } = this.props;
        const { verticalAlign } = options || {};
        return (cellData: GridCellData): React.ReactNode => {
            const { colStartIndex, colEndIndex, rowStartIndex, rowEndIndex } = cellData;
            return (
                <GridCell
                    colStartIndex={colStartIndex}
                    colEndIndex={colEndIndex}
                    rowStartIndex={verticalAlign === "bottom" && rowEndIndex ? rowEndIndex - 1 : rowStartIndex}
                    rowEndIndex={verticalAlign === "top" || verticalAlign === "bottom" ? undefined : rowEndIndex}
                    key={`${colStartIndex}-${rowStartIndex}`}
                >
                    {this.renderFlowElement(cellData)}
                </GridCell>
            );
        };
    };

    render(): React.ReactElement {
        const { flow, options } = this.props;
        const verticalModelAlign = options && options.verticalAlign === "bottom" ? "bottom" : "top";
        const { gridCellData, columnCount } = buildRenderData(flow, verticalModelAlign);
        return (
            <div className="flow-modeler" onClick={this.clearSelection} style={{ gridTemplateColumns: `repeat(${columnCount}, max-content)` }}>
                {gridCellData.map(this.renderGridCell())}
            </div>
        );
    }

    static propTypes = {
        flow: PropTypes.shape({
            firstElementId: PropTypes.string.isRequired,
            elements: PropTypes.objectOf(
                PropTypes.oneOfType([
                    PropTypes.shape({
                        data: PropTypes.object,
                        nextElementId: PropTypes.string
                    }),
                    PropTypes.shape({
                        data: PropTypes.object,
                        nextElements: PropTypes.arrayOf(
                            PropTypes.shape({
                                id: PropTypes.string.isRequired,
                                conditionData: PropTypes.object.isRequired
                            })
                        ).isRequired
                    })
                ])
            ).isRequired
        }).isRequired,
        options: PropTypes.shape({
            verticalAlign: PropTypes.oneOf(["top", "middle", "bottom"]),
            readOnly: PropTypes.bool
        }),
        renderContent: PropTypes.func.isRequired,
        renderGatewayConditionType: PropTypes.func,
        renderGatewayConditionValue: PropTypes.func
    };
}
