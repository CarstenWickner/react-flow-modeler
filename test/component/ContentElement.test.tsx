import * as React from "react";
import { shallow } from "enzyme";

import { ContentElement } from "../../src/component/ContentElement";

describe("renders correctly", () => {
    const onSelect = (): void => {};
    it("with minimal props", () => {
        const component = shallow(
            <ContentElement elementId="element-id" editMenu={<div className="edit-menu-placeholder" />} onSelect={onSelect}>
                {"text"}
            </ContentElement>
        );
        expect(component).toMatchSnapshot();
    });
    it("when not selected", () => {
        const component = shallow(
            <ContentElement elementId="element-id" editMenu={undefined} onSelect={onSelect}>
                {"text"}
            </ContentElement>
        );
        const contentElement = component.find(".content-element");
        expect(contentElement.hasClass("selected")).toBe(false);
        expect(component.find(".edit-menu-placeholder").exists()).toBe(false);
    });
});
describe("calls onSelect", () => {
    const onSelect = jest.fn(() => {});
    const event = ({ stopPropagation: jest.fn(() => {}) } as unknown) as React.MouseEvent;
    it("on click event", () => {
        const component = shallow(
            <ContentElement elementId={"element-id"} editMenu={undefined} onSelect={onSelect}>
                {"text"}
            </ContentElement>
        );
        component.find(".content-element").prop("onClick")(event);
        expect(onSelect.mock.calls).toHaveLength(1);
        expect(onSelect.mock.calls[0]).toHaveLength(1);
        expect(onSelect.mock.calls[0][0]).toEqual("element-id");
    });
});
