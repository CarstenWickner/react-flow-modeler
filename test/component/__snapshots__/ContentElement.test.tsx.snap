// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with minimal props 1`] = `
<Fragment>
  <div
    className="stroke-horizontal arrow"
  />
  <div
    className="flow-element content-element selected"
    onClick={[Function]}
  >
    text
  </div>
  <div
    className="edit-menu-placeholder"
  />
  <HorizontalStroke
    optional={true}
  />
</Fragment>
`;
